qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,359 @@
+    def supported_basis(self):
+        """Returns a dictionary of supported basis for synthesis
+
+        This is expected to return a dictionary where the key is a string
+        basis and the value is a list of gate names that the basis works in.
+        If the synthesis method doesn't support multiple basis this should
+        return ``None``.
+
+        If a dictionary is returned by this method the run kwargs will be
+        passed a parameter ``matched_basis`` which contains a list of the
+        basis strings which match the target basis gate set for the
+        transpilation.
+        """
+        pass
+
+    @abc.abstractmethod
+    def run(self, unitary, **options):
+        """Run synthesis for the given unitary matrix
+
+        Args:
+            unitary (numpy.ndarray): The unitary matrix to synthesize to a
+                :class:`~qiskit.dagcircuit.DAGCircuit` object
+            options: The optional kwargs that are passed based on the output
+                of :meth:`supports_basis_gates`, :meth:`supports_coupling_map`,
+                and :meth:`supports_approximation_degree`. If
+                :meth:`supports_coupling_map` returns ``True`` a kwarg
+                ``coupling_map`` will be passed either containing ``None`` (if
+                there is no coupling map) or a
+                :class:`~qiskit.transpiler.CouplingMap` object. If
+                :meth:`supports_basis_gates` returns ``True`` then a kwarg
+                ``basis_gates`` will the list of basis gate names will be
+                passed. Finally if :meth:`supports_approximation_degree`
+                returns ``True`` a kwarg ``approximation_degree`` containing
+                a float for the approximation value will be passed.
+
+        Returns:
+            DAGCircuit: The dag circuit representation of the unitary. Alternatively,
+                you can return a tuple of the form ``(dag, wires)`` where ``dag`` is the dag
+                circuit representation of the circuit representation of the unitary
+                and ``wires`` is the mapping wires to use for
+                :meth:`qiskit.dagcircuit.DAGCircuit.substitute_node_with_dag`.
+        """
+        pass
+
+
+class UnitarySynthesisPluginManager:
+    """Unitary Synthesis plugin manager class
+
+    This class tracks the installed plugins, it has a single property,
+    ``ext_plugins`` which contains a list of stevedore plugin objects.
+    """
+
+    def __init__(self):
+        self.ext_plugins = stevedore.ExtensionManager(
+            "qiskit.unitary_synthesis", invoke_on_load=True, propagate_map_exceptions=True
+        )
+
+
+def unitary_synthesis_plugin_names():
+    """Return a list of installed unitary synthesis plugin names
+
+    Returns:
+        list: A list of the installed unitary synthesis plugin names. The
+            plugin names are valid values for the
+            :func:`~qiskit.compiler.transpile` kwarg
+            ``unitary_synthesis_method``.
```

Comment:
```suggestion
        list: A list of the installed unitary synthesis plugin names. The plugin names are valid
        values for the :func:`~qiskit.compiler.transpile` kwarg ``unitary_synthesis_method``.
```
similar continuation paragraph